doc/sphinx/proof-engine/ssreflect-proof-language.rst

```diff
@@ -2735,7 +2691,7 @@ type classes inference.
   Full inference for ``ty``. The first subgoal demands a
   proof of such instantiated statement.
 
-+ coqtop::
++ .. coqdoc::
```

Comment:
Possibly. Does it work in a plain `coqtop`?

Reply:
No it gives the same error.

Reply:
OK. Then it's not a Sphinx issue, right?

Reply:
Hum, why do you say this?

Reply:
Sorry, maybe I misunderstood your comment. I thought you meant that coqtop complained about that input as well.

I think we misunderstood each other. By _does it work in a plain coqtop_, I meant "does it work if you start a coqtop process on the command line and paste the buggy input".  But now I realize (I think) that you understood it as "does it work if you use a plain coqtop directive" in Sphinx.

Phew :)

My best guess would be that coqtop doesn't understand this inpput as a full sentence, so it hangs waiting for more input, and sphinx hangs waiting for output, yielding the timeout that you observe).  This is a problem that happens commonly with PG.  Are you sure that this sentence is complete, and does coqtop respond to it immediately?

Reply:
Oh sorry I should have explicited from the beginning that this is a full correct sentence (I tested it in CoqIDE only).

Reply:
Oh sorry I should have explicited from the beginning that this is a full correct sentence (I tested it in CoqIDE only).

Reply:
FTR I just tried this with `coqtop`:

```
From Coq Require Import ssreflect.
Axiom ty : Type.
Goal True.
have foo : ty := .
```

There is indeed something fishy going on: it does work but it doesn't print the resulting subgoals... (it prints nothing, but using `Show` then allows to see the two subgoals).
In CoqIDE, it prints the two subgoals immediately.
@gares @ejgallego Any idea of what might be going on?

Reply:
I find this kind of post dismissive and hurtful.

> It is not perfect but the old one was much worse [involved messy static analysis].

The new solution may be better in terms of implementation simplicity on the Coq side.  It isn't clear whether that's the right metric here.  Breaking IDEs should have to clear a higher bar than making the implementation prettier.

> I have been saying this for a while, coqtop is meant for humans and not for programmatic use, but people do still insist on using it.

Can you suggest a more appropriate solution? I'd love to know what the right way to do things here is.

Assuming good faith goes long ways. "people" (in this case, me) don't wrap coqtop wrappers for fun.

> Goal printing should be always IDE-initiated.

There's no IDE here, so I'm not sure what this means.

> A more robust solution here would be to call coqtop -quiet and have sphinx do a Show indeed when it is expecting goals.

This doesn't make sense to me.  Sphinx doesn't know when to expect a goal.

(The way the manual has always worked is that it printed the output of Coqtop for each line of input.  IOW, the manual has always assumed coqtop: that's the Coq interface that it's demoing.)

Reply:
So indeed `coqtop` does fully understand the sentence, it is just it considers the command a "control command" [such as `Set Printing Foo`] and doesn't' reprint the goals.

Reply:
@zimmi48 can you try adding `Show.` on the same line?

Reply:
So a short summary, in both old and new versions the only reliable way to get goals for `coqtop` clients is to use `-quiet` and explicitly ask for them.

Reply:
Shouldn't there be a `-verbose` option that prints every time? Since `Show` indeed only works inside a proof.
BTW, in this particular case, the "heuristic" seem broken wrt the property you gave since the goals do change.

Reply:
Oh indeed maybe the heuristic is broken, I wouldn't be surprised, let me have a look.

Reply:
> Shouldn't there be a -verbose option that prints every time? Since Show indeed only works inside a proof.

Yup, we should add something. You have `-load-verbose`, but we are also using Time now which is far from optimal, and as we all know not very robust due to the output format.

Reply:
Umm, I cannot see a problem with the goal equality here, https://github.com/coq/coq/blob/b74d9500e5943317f1baf4f36b3d979d40f6105f/toplevel/coqloop.ml#L330 , @Zimmi48 you know more about the `proof` object, why are they considered equal?

Reply:
> There is indeed something fishy going on: it does work but it doesn't print the resulting subgoals... 

Printing nothing should work fine, so there's still quite possibly a bug on the Sphinx side. (There are plenty of commands that don't print anything, as you pointed emilio).  Sphinx and PG get stuck when they think they sent a full sentence but Coq doesn't agree.

I'll look at this tonight.

Reply:
> … but Sphinx still shouldn't crash (it should see the header and understand that there was no output). Will investigate tonight.

Funnily enough this looks like due to some grammar change, not the toplevel [we would still need to bisect to be 100% sure], but indeed I don't Sphinx is at fault here; indeed coqtop is stuck at parsing so that is hard to avoid without introducing an end-of-command separator.

The "spec" of coqtop says that after each command there should be at least a newline, I consider that stable enough as for Sphinx to rely on that.

Reply:
I think the workaround of using `.. coqdoc::` is actually fine but yes, please open a bug about this (this PR has already received too many comments about this issue).

@cpitclaudel Another issue in passing which discourages me from using `.. coqdoc::` more often in place of `.. coqtop:: in undo` is that the rendering is not exactly the same (in particular the margins).

Reply:
Yeah, but the hanging on parsing issue still persists and I don't see any other way to avoid it than by adding a command terminator. Actually, I am not sure we could legitimate claim than there is a bug in the grammar, the bug is in the coqtop input model I think; so I'll wait for @cpitclaudel opinion before proceeding.

If the solution we choose is to add a command terminator I do prefer to directly submit a PR implementing it instead of an issue.

Reply:
Note that detecting the header is not going to work, as it is printer by the lexer!!!!

So yes, while the lexer is waiting for more input, if it detects an EOL will set a flag so the next call prints a header.

Reply:
> I do prefer to directly submit a PR implementing it instead of an issue.

Opening an issue first is always better to clearly describe the problem encountered to others.

> If the solution we choose is to add a command terminator I do prefer to directly submit a PR implementing it instead of an issue.

This looks more like a workaround than a fix to me.

Reply:
Ok, I'll open an issue, but it will take me some time as it is going to be a large issue to describe, so no ETA.

> This looks more like a workaround than a fix to me.

Umm, I'm pretty sure that's the proper fix; a client should signal the end of input; in the case of `coqtop` the only supported way is to close the input stream, but then you are stuck.